### ISSUE TITLE
Preserve `karabiner.json` symlink while writing?

### ISSUE THREAD

**Issue by lgarron** [pqrs-org/Karabiner-Elements#3248]
It's a pretty common pattern to track dotfiles in a `git` repository and symlink all the files, to the point that there's a whole [ecosystem](https://dotfiles.github.io/) around this. Unfortunately, if you symlink `karabiner.json` into `~/.config/karabiner/karabiner.json`, then Karabiner-Elements will overwrite it, instead of editing the symlinked file. This is different from a lot of other programs. At best, this causes noticable errors when using programs like `stow`. At worst, this makes it easy to silently miss Karabiner config changes in one's tracking repository.

This behaviour seems to be implemented in `async_file_writer.hpp`:

https://github.com/pqrs-org/Karabiner-Elements/blob/803b642ee4f1203c91bdbe9025e26b378bef571d/src/share/async_file_writer.hpp#L36

That code was originally introduced in https://github.com/pqrs-org/Karabiner-Elements/commit/973d33be9c6785b32af9b812a5387b6a1cd35499 , without further context.

Would it be possible to either:

- replace the contents of `karabiner.json` without first unlinking it, or
- offer a setting to do so?

This would make Karabiner a lot more convenient to use, for those of us with multiple computers!

**Comment by iamtodor** [comment]
+1 for this feature. I tried to keep `karabiner.json` in dotfiles repo and met the same logic: once I modify any keybind karabiner overwrites the `karabiner.json` instead of modifying symlink

**Comment by teddywing** [comment]
The documentation describes that it is possible to symlink the configuration file, but the entire `~/.config/karabiner` directory must be symlinked instead of the “karabiner.json” file:

> If you want to move karabiner.json to another place and make symbolic link, make a symbolic link to ~/.config/karabiner directory instead of karabiner.json.
>
> Warning
> 
> Do not make a symlink to karabiner.json directly.
> 
> Karabiner-Elements will fail to detect the configuration file update and fail to reload the configuration if karabiner.json is a symbolic link.

(https://karabiner-elements.pqrs.org/docs/manual/misc/configuration-file-path/#about-symbolic-link)

I’m not clear why “karabiner.json” can’t be directly symlinked. If I had to guess, maybe it’s due to the way the file watcher is set up.

**Comment by iamtodor** [comment]
@teddywing thank you for sharing your findings!

In `.config/karabiner` a few dirs:

```
exa ~/.config/karabiner
Permissions Size User            Date Modified Name
drwx------     - user 18 Jul 11:57  assets/
drwx------     - user 17 Oct 08:15  automatic_backups/
.rw-------  8.7k user 17 Oct 08:24  karabiner.json
```

I'm not sure why would I want to keep backups or assets under my dotfiles repo. It's quite possible I'm missing something. I'd be glad if you could provide your thoughts or guidance here :)

**Comment by teddywing** [comment]
> I'm not sure why would I want to keep backups or assets under my dotfiles repo. There is a chance, that I'm missing something. I'd be glad if you could provide your thoughts or guidance here :)

As far as I know, only “karabiner.json” is needed to save your configuration. That’s the only file I track from `~/.config/karabiner` in my dotfiles and I gitignore the rest.

**Comment by iamtodor** [comment]
@teddywing 
> As far as I know, only “karabiner.json” is needed to save your configuration. That’s the only file I track from `~/.config/karabiner` in my dotfiles and I gitignore the rest.

Yeah, exactly and this is the point of having `karabiner.json` separately from the backups.

> That’s the only file I track from ~/.config/karabiner in my dotfiles and I gitignore the rest.

Then the next time you set up your new machine means that you need to create `assets` and `automatic_backups` manually or move these folders manually to the dotfiles dir, right? Otherwise, I don't understand how can you manage to have 
`karabiner.json` in your dotfile and link it properly on the new/wiped machine.

**Comment by teddywing** [comment]
> Then the next time you set up your new machine means that you need to create `assets` and `automatic_backups` manually or move these folders manually to the dotfiles dir, right?

Karabiner automatically creates all of the files and folders in `~/.config/karabiner` (including “karabiner.json”) if they’re not present, so there’s no need to manually create them or store the additional folders in your dotfiles.

That said, nothing prevents you from tracking everything under `~/.config/karabiner` in your dotfiles if that’s what you prefer.

I haven’t downloaded any complex modifications so I’m not sure, but it’s possible those are stored in `assets/` and that you’d need to track those too.

**Comment by iamtodor** [comment]
@teddywing 

> Karabiner automatically creates all of the files and folders in ~/.config/karabiner (including “karabiner.json”) if they’re not present, so there’s no need to manually create them or store the additional folders in your dotfiles

Yeah, and I assume it would not allow me to automatically create a symlink. Perhaps, I need to test it out.

> I haven’t downloaded any complex modifications

Same here. Just want to store a config file properly :D

**Comment by teddywing** [comment]
> Yeah, and I assume it would not allow me to automatically create a symlink.

I’m not clear what would prevent you from creating the symlink. If the `~/.config/karabiner` folder already exists, you can remove the automatically generated folder and replace it with your symlink.

**Comment by ianrobrien** [comment]
This is also an issue when using [chezmoi](https://www.chezmoi.io) as it [cannot create symbolic links to entire directories](https://www.chezmoi.io/user-guide/frequently-asked-questions/design/#what-are-the-limitations-of-chezmois-symlink-mode).

**Comment by qd3v** [comment]
Yup, my favourite [rcup](https://github.com/thoughtbot/rcm) works on file-level too...

**Comment by ssbarnea** [comment]
This also prevents using [stow](https://www.gnu.org/software/stow/) to keep karabiner config in .dotfiles, as it cannot symlink folders, only individual files.

**Comment by iammohitsakhuja** [comment]
> This also prevents using [stow](https://www.gnu.org/software/stow/) to keep karabiner config in .dotfiles, as it cannot symlink folders, only individual files.

Exactly this! Most folks use the `--no-folding` option of `stow` to keep their dotfiles clean from auto-generated files created by various programs. Some programs go even further and ignore common `XDG_` patterns, dumping everything into `~/.config/<program_name>`. This makes it impractical to use `stow` without `--no-folding`. 

The problem is that with `--no-folding`, `stow` creates a symlink for the underlying file rather than the folder itself, but Karabiner overwrites that symlink whenever it updates `karabiner.json`.
